### ISSUE TITLE
[bug]: vertical separator not rendering

### ISSUE THREAD

**Issue by OliverGilan** [shadcn-ui/ui#4818]
### Describe the bug

When using the separator with a "vertical" orientation it does not render properly. Looking at the browser dev tools it appears to have a size of 1x0 pixels. 

Changing the default style here to use `min-h-full` instead of just `h-full` seems to fix this.

Seems to be the same as this [issue](https://github.com/huntabyte/shadcn-svelte/issues/1149) encountered in the shadcn-svelte port

### Affected component/components

Separator

### How to reproduce

1. Place a Separator in a flex parent with two other components and orientation "vertical"
```javascript
<div className="flex flex-row flex-nowrap gap-4 h-fit">
          <p>Foo</p>
          <Separator orientation="vertical" />
          <p>Bar</p>
 </div>
```
2. The separator will not appear
3. Change style within components/ui/separator.tsx
```javascript
<SeparatorPrimitive.Root
      ref={ref}
      decorative={decorative}
      orientation={orientation}
      className={cn(
        'shrink-0 bg-border',
        orientation === 'horizontal' ? 'h-[1px] w-full' : 'min-h-full w-[1px]',
        className
      )}
      {...props}
    />
```
4. Separator will now render

### Codesandbox/StackBlitz link

https://stackblitz.com/edit/stackblitz-starters-t7h9sc?file=components%2Fui%2Fseparator.tsx

### System Info

```bash
M2 Mac,
Firefox
```

### Before submitting

- [X] I've made research efforts and searched the documentation
- [X] I've searched for existing issues

**Comment by OliverGilan** [comment]
Opened a PR to help fix this!

**Comment by TharindaPrabhath** [comment]
In my case, I had to define the height explicitly in the className prop. Then, it was displayed.

**Comment by OliverGilan** [comment]
Yes explicitly setting a height does work but the default behavior of the component should be that it fills up the height of the parent which isn't happening. Can always override that with the className prop even with this change

**Comment by TharindaPrabhath** [comment]
Yes, agree.

**Comment by gerardsiles** [comment]
it has an easy fix, on the primitive root change:

     className={cn(
        "shrink-0 bg-border",
        orientation === "horizontal" ? "h-[1px] w-full" : "full w-[1px]",
        className,
      )}

      for:

           className={cn(
        "shrink-0 bg-border",
        orientation === "horizontal" ? "h-[1px] w-full" : "min-h-full w-[1px]",
        className,
      )}

setting min-h-full should fix this this issue when set to vertical

**Comment by OliverGilan** [comment]
Yes that is the fix I used and it's what I have in the PR I opened for this

**Comment by KarlG-03** [comment]
doesn't seem to work if the parent is display `flex` and has `items-center` to center fields

```
<div className="flex space-between items-center">
                <div className="flex-1 text-sm text-center">
                  Location
                </div>
                <Separator orientation="vertical" className="h-full" />
                <div className="flex-1 text-sm text-center">
                  Color
                </div>
                <Separator orientation="vertical" />
                <div className="flex-1 text-sm text-center">
                  Size
                </div>
</div>
```

had to go for grid
```
<div
                className="grid"
                style={{ gridTemplateColumns: '1fr 1px 1fr 1px 1fr' }}
              >
                <div className="text-sm text-center">
                  Location
                </div>
                <Separator orientation="vertical" />
                <div className="text-center">
                  Color
                </div>
                <Separator orientation="vertical" />
                <div className="text-center">
                  Size
                </div>
              </div>
              <div>
```

**Comment by deilon** [comment]
The `orientation="vertical"` will not show if you add class `items-center` to its parent element.

**Comment by sheikhameen** [comment]
When `orientation="vertical"`, you can simply set the Separator's height to auto: `h-auto`. Doesn't work when parent has `items-center` tho.

**Comment by ld-bravo** [comment]
I had to use min-h-[] and border:

```html
<div className="flex items-center">
          <Separator orientation="vertical" className="min-h-[35px] m-[5px] border-[1px]" />
</div>
```

**Comment by jcajuab** [comment]
I was using the block [sidebar-08](https://ui.shadcn.com/view/styles/new-york/sidebar-08) and encountered the same issue rendering the new vertical separator with the provided code:

```tsx
<div className="flex items-center gap-2 px-4">
  <Separator orientation="vertical" className="mr-2 h-4" />
</div>
```

I was able to fix it by updating the class like this:

```tsx
<div className="flex items-center gap-2 px-4">
  <Separator orientation="vertical" className="mr-2 data-[orientation=vertical]:h-4" />
</div>
```

Hope this helps!

**Comment by Joshua-hypt** [comment]
Still open

**Comment by garlic-brewlabs** [comment]
I was unable to have success using `h-full` on the parent element, as suggested by @JoshuaS98 

What did work (in my case) was `<Separator orientation="vertical" className="data-[orientation=vertical]:h-4" />` as suggested by @jcajuab 

Replication here https://bolt.new/~/sb1-pvlzdufr (see hero component there is a separator between the buttons)

**Comment by Joshua-hypt** [comment]
@garlic-brewlabs That's because your container depends on the height of it's inner content. Give the parent container an absolute height (like `h-12`) and it will be fixed. I literally quoted:

**If the height of the containing block is not specified explicitly (i.e., it depends on content height), and this element is not absolutely positioned, the value computes to auto**

**Comment by jjlinares** [comment]
I ran into this same issue and the parent component needs a height set which sucks if you have a dynamic height component (my case).

But most likely you don't need the Separator component, all you need is a border and some padding.

**Comment by tushar1998** [comment]
Its not possible to fix directly in shadcn there are various scenarios where it cannot work all cannot be covered. Another issue is this below wont work 
```
<div className="border border-gray-300 rounded-md p-2 flex h-full gap-2">
    <div className="flex-1 p-2">Hello</div>
    <Separator orientation="vertical" className="h-auto" />
    <div className="flex-1 p-2">World</div>
</div>
```
but this below works
```
<div className="border border-gray-300 rounded-md p-2 flex h-full gap-2">
    <div className="flex-1 p-2">Hello</div>
    <Separator orientation="vertical" className="data-[orientation=vertical]:h-auto" />
    <div className="flex-1 p-2">World</div>
</div>
```

There is an issue with tailwind-merge it cannot identify custom state based dynamic styles. We need to manually check scenarios and override appropriately. Current implementation of Seperator is just fine

**Comment by sglza** [comment]
`self-stretch` seems to work very well

```tsx
className={cn(
  'shrink-0 bg-border-divider data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:w-px data-[orientation=vertical]:self-stretch',
  className
)}
```

**Comment by maxktz** [comment]
add `self-stretch` to the separator is the top solution

**Comment by MazenChaouch** [comment]
I think the best fix is to update the component classes so the vertical orientation uses `h-auto` instead of `h-full`, and also apply `self-stretch`. 

Current:
```typescript 
className={cn(
  "bg-border shrink-0 data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-px",
  className,
)}
``` 

Proposed update:
```typescript 
className={cn(
  "bg-border shrink-0 data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-auto data-[orientation=vertical]:w-px data-[orientation=vertical]:self-stretch",
  className,
)}
``` 

This adjustment makes the vertical separator behave correctly in all cases.

**Comment by sglza** [comment]
in what cases did you find `h-auto` useful? I could't find any at the time

**Comment by MilanObrenovic** [comment]
> I think the best fix is to update the component classes so the vertical orientation uses `h-auto` instead of `h-full`, and also apply `self-stretch`.
> 
> Current:
> 
> className={cn(
>   "bg-border shrink-0 data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-full data-[orientation=vertical]:w-px",
>   className,
> )}
> Proposed update:
> 
> className={cn(
>   "bg-border shrink-0 data-[orientation=horizontal]:h-px data-[orientation=horizontal]:w-full data-[orientation=vertical]:h-auto data-[orientation=vertical]:w-px data-[orientation=vertical]:self-stretch",
>   className,
> )}
> This adjustment makes the vertical separator behave correctly in all cases.

This fixed it for me. thanks

**Comment by glukki** [comment]
To make Separator work both in a container with a fixed height, and with noi height defined, it has to have both `min-height: 100%` and `align-self: stretch;`.

Or in tailwind terms: `data-[orientation=vertical]:min-h-full data-[orientation=vertical]:self-stretch`

Full class:
```js
"shrink-0 bg-border data-[orientation=horizontal]:h-px data-[orientation=vertical]:min-h-full data-[orientation=horizontal]:w-full data-[orientation=vertical]:w-px data-[orientation=vertical]:self-stretch"
```


Cases for your Storybook
```tsx

/**
 * Use `orientation="vertical"` for a vertical separator between horizontal items in a container with a fixed height.
 */
export const VerticalWithFixedContainerHeight: Story = {
    render: () => (
        <div className="flex h-12 items-center justify-center gap-2">
            <div>Left</div>
            <Separator orientation="vertical" />
            <div>Right</div>
        </div>
    ),
};

/**
 * Use `orientation="vertical"` for a vertical separator between horizontal items in a container without a fixed height.
 */
export const VerticalWithUnsetContainerHeight: Story = {
    render: () => (
        <div className="flex items-center justify-center gap-2">
            <div>Left</div>
            <Separator orientation="vertical" />
            <div>Right</div>
        </div>
    ),
};
```

<img width="834" height="499" alt="Image" src="https://github.com/user-attachments/assets/9741d234-21da-4566-bde1-4a130187c590" />

**Comment by JamesDennisOrg** [comment]
Having tried all of the previous suggested fixes, and none of them working in Next.js 16, Tailwind v4 - simply adding : className="py-2" - "py-3" - "py-4" etc... fixes the rendering. Higher padding value = taller vertical separator.

(add it to the native styles in components/ui/separator - or just add a class to the element in the jsx)

**Comment by basitol** [comment]
Opened a fix for this: #11813. Turns out min-h-full (suggested above) has the same underlying issue since percentage min-height also needs a definite containing block height 

used self-stretch instead.

[https://github.com/shadcn-ui/ui/pull/11813](url)

**Comment by scenerecap** [comment]
Checked this on ui.shadcn.com today (2026-09-16, Chromium 151, 1280×720) — it no longer reproduces for me on the documentation page:

1. Opened https://ui.shadcn.com/docs/components/base/separator
2. Scrolled to the **Vertical** section
3. The demo renders "Blog / Docs / Source" with two vertical separators inside a `flex items-center` row — no explicit height on the parent container
4. Both separators are fully visible (~20 px tall), rendering correctly via `align-self: stretch`

The deployed component uses `data-vertical:self-stretch` (Tailwind v4 syntax) in place of `h-full`, so `height: 100%` no longer resolves to zero in flex parents without a fixed height.

Recording of the current docs-page behaviour: https://scenerecap.com/s/e8zwf6d53b

PR #11813 is still open and targets the CLI-generated `separator.tsx` template — whether that generated file also needs the same change is a separate question for maintainers.
